### ISSUE TITLE
No way to ensure line existence with file.line in most circumstances

### ISSUE THREAD

**Issue by monolied** [saltstack/salt#58587]
### Description of Issue
file.line should provide a way to ensure the existence of a line without strict conditions. But because of some semantic bugs and issues, it doesn't.

### Setup
I need an entry in the /etc/hosts of some systems. It has to be there; it's entirely unimportant, where exactly it is in the file.
(I know I can use file.append in this special case where I'm fine with putting it at the end of a file. Let's for the sake of the argument assume I have to put it somewhere else. In those cases, I'd try file.line, but fail. A lot.)

### Steps to Reproduce Issue

hosts file with at least two matching strings which I will use 'this-is-the-pattern' for. eg:

```
127.0.0.1       localhost
0.0.0.0    this-is-the-pattern
10.0.0.1    testhost.domain.de testhost
0.0.0.0    this-is-the-pattern
```
sls:
```
modify_resolver:
   file.line:
    - name: /etc/hosts
    - mode: ensure
    # - mode: insert
    - content: 10.1.1.1    added-by-salt.domain.de
    - after: .*this-is-the-pattern
    # - location: end
```
If I use "ensure":
   - I have to use "before" and/or "after"
   - If I use "after" I get an error message because it matches twice (`"Found more than expected occurrences in "after" expression"`).

So I use "insert", because it is supposed to ignore multiple matches.
   - Again I have to use "before" and/or "after" (and/or "location"). 
  - So I use "after", but again get an error message about multiple locations of the after string.  Alright, misunderstood the documentation part: "multiple (and non-matching) lines are allowed between before and after, if they are specified"). Alas, it doesn't allow for multiple lines matching the pattern (which *is* documented for the "before" and "after" tags. However, why not allow multiple matches and add after the last/before the first? But I digress.)

I still got hope, though:
  - "insert" also offers "location" instead of the fickle "after". So I try "location: end". But now it ignores the fact the string already exists and inserts it at the end anyway! Not only I have it double, although "insert" supposedly only inserts, if the line doesn't exist already, I also risk of an endless insertion of strings if I use two of those states on the same file.

In summary, an option to plain ensure the existence of a line in a file and adding it in a certain position *not* being EOF in case it doesn't exist which works in very simple circumstances (file without guaranteed *and* singular existence of a matchstring) is direly missing, and/or some of the documentation is misleading, and/or there's some bugs with the interpretation of the file.line state.

### Versions Report

Salt Version:
           Salt: 3000.3

Dependency Versions:
           cffi: Not Installed
       cherrypy: unknown
       dateutil: 2.7.3
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
         Jinja2: 2.10
        libgit2: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.6.2
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.16 (default, Oct 10 2019, 22:02:15)
   python-gnupg: Not Installed
         PyYAML: 3.13
          PyZMQ: 17.1.2
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.3.1

System Versions:
           dist: debian 10.6
         locale: UTF-8
        machine: x86_64
        release: 4.19.0-11-amd64
         system: Linux
        version: debian 10.6

**Comment by garethgreenaway** [comment]
@monolied Thanks for the report.  If you're looking to manage the hosts file on systems have you considered using the `hosts` state module? https://docs.saltstack.com/en/latest/ref/states/all/salt.states.host.html#salt.states.host.present

**Comment by monolied** [comment]
Thanks for the hint. I'm aware of the hosts state module, and it would certainly solve the immediate issue in this case. I tried to use file.line because it promised to allow for some control over the order of items, to improve organization, readability and semantic issues (with the hosts file, order of items can be relevant). And the shortcomings of file.line certainly warrant an issue report. That aside, thanks again for your help.

**Comment by craigafinch** [comment]
@garethgreenaway I have run into numerous other configuration files in which I have faced this exact problem. We can't create a custom Salt module for every application config file. It's a really common use case to ensure that a specific line is present (but not duplicated) in specific location in a text file. I have struggled with using file.replace and some complex regular expressions to achieve the simple result of ensuring that a line is present (and doesn't get re-added every time the state runs)

**Comment by mm-irvingleonard** [comment]
@monolied There's a workaround while they fix this state module: assuming that you're using a regex for your mapping (something as simple as an extra space might break your string comparison), you have `grep` available, and you're running a salt version recent enough then you can set an `unless` state requirement and populate it with `grep -Pxq "\s*0\.0\.0\.0\s+this-is-the-pattern" /etc/hosts` (keep in mind that `.` has a special meaning in PCRE, and also `-` in certain conditions). Basically you rely in grep to do the checking for you; if the pattern is found grep returns 0 and nothing else happens, otherwise the state is run and your line gets added (following all the state logic: ensure, insert, before, after, location, etc.).

**Comment by mm-irvingleonard** [comment]
@garethgreenaway Assuming you search for the string early in the process, a match should stop any further processing in the `ensure` and `insert` modes, we're in the right state, nothing to do. It should be the opposite for the 'delete' mode, where the absence of a match should stop any further processing, it's already deleted, nothing to do. And then there's the rabbit hole (the `replace` mode).

This logic would be about the content of the file, obviously there's more processing related to the metadata that needs to happen: check for owner, group, file_mode.

There's also a simple bug there: if you enable the `create` switch the state will fail in a test run even though it will succeed on a regular run, go figure.

**Comment by garethgreenaway** [comment]
Closing this due to age and the old version of Salt.
Can you retest this with the latest release of Salt 3006.2 and if this is still an issue, please open a new issue.
